File: python/core/src/smooth_operator_core/extension/protocol.py
"""SEP wire protocol — JSON-RPC 2.0 frames and typed method params/results.

SEP (the Smooth Extension Protocol) is JSON-RPC 2.0 over ndjson on an extension
subprocess's stdio. The canonical schemas live in the ``smooth-operator`` repo at
``spec/extension/``; the types here are the Python host's view of that wire. Field
names are ``snake_case`` to match the spec exactly.

The Python sibling of the Rust reference ``protocol.rs``. The host works mostly in
plain dicts (the wire is JSON); the dataclasses here are the typed views the
handshake, hook, and tool paths parse into — and what the conformance suite
replays to catch drift from the wire.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Optional

#: The SEP protocol version this host implements.
PROTOCOL_VERSION = 1


class codes:
    """JSON-RPC error codes: the standard range plus the SEP extensions documented
    in ``spec/extension/envelope.md``."""

    PARSE_ERROR = -32700
    INVALID_REQUEST = -32600
    METHOD_NOT_FOUND = -32601
    INVALID_PARAMS = -32602
    INTERNAL_ERROR = -32603

    #: A hook or policy vetoed the operation.
    BLOCKED = -32000
    #: ``ui/request`` in a headless/uncapable frontend.
    NO_UI = -32001
    #: Extension acted beyond its granted trust.
    NOT_TRUSTED = -32002
    #: Command-tier action attempted from an event-tier context.
    CONTEXT_VIOLATION = -32003
    #: Method requires a capability the handshake did not enable.
    CAPABILITY_DISABLED = -32004
    #: Request cancelled via ``$/cancel``.
    CANCELLED = -32800


class method:
    """SEP method names, centralized so the host and tests never spell one wrong."""

    INITIALIZE = "initialize"
    SHUTDOWN = "shutdown"
    PING = "ping"
    EVENT = "event"
    HOOK = "hook"
    TOOL_EXECUTE = "tool/execute"
    TOOL_UPDATE = "tool/update"
    COMMAND_EXECUTE = "command/execute"
    COMMAND_COMPLETE = "command/complete"
    CANCEL = "$/cancel"
    REGISTRY_UPDATE = "registry/update"
    TOOLS_SET_ACTIVE = "tools/set_active"
    EXEC_RUN = "exec/run"
    UI_REQUEST = "ui/request"
    LOG = "log"
    BUS_PUBLISH = "bus/publish"
    SESSION_SEND_MESSAGE = "session/send_message"
    SESSION_SEND_USER_MESSAGE = "session/send_user_message"
    SESSION_APPEND_ENTRY = "session/append_entry"


class RpcError(Exception):
    """A JSON-RPC error object, raised across the ext<->host seam."""

    def __init__(self, code: int, message: str, data: Any | None = None) -> None:
        super().__init__(f"JSON-RPC error {code}: {message}")
        self.code = code
        self.message = message
        self.data = data

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"code": self.code, "message": self.message}
        if self.data is not None:
            out["data"] = self.data
        return out

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> RpcError:
        return cls(int(d["code"]), str(d["message"]), d.get("data"))

    def __eq__(self, other: object) -> bool:
        return (
            isinstance(other, RpcError)
            and other.code == self.code
            and other.message == self.message
            and other.data == self.data
        )


class Tier(str, Enum):
    """Whether a dispatch may only observe (``EVENT``) or may mutate the session
    (``COMMAND``). Session-mutating ext->host actions require ``COMMAND``."""

    EVENT = "event"
    COMMAND = "command"


@dataclass
class Context:
    """The dispatch context carried by every host->ext event/hook/tool/command."""

    token: str
    tier: Tier

    def to_dict(self) -> dict[str, Any]:
        return {"token": self.token, "tier": self.tier.value}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> Context:
        return cls(token=str(d["token"]), tier=Tier(d["tier"]))


# ---------------------------------------------------------------------------
# The JSON-RPC envelope.
# ---------------------------------------------------------------------------


@dataclass
class Message:
    """The JSON-RPC 2.0 envelope. All four frame shapes share this dataclass; which
    fields are present determines the shape:

    - request: ``id`` + ``method`` (+ optional ``params``)
    - notification: ``method``, no ``id``
    - success response: ``id`` + ``result``
    - error response: ``id`` + ``error``
    """

    jsonrpc: str = "2.0"
    id: Any | None = None
    method: Optional[str] = None
    params: Any | None = None
    result: Any | None = None
    error: Optional[RpcError] = None

    @classmethod
    def request(cls, id: Any, method: str, params: Any) -> Message:
        return cls(id=id, method=method, params=params)

    @classmethod
    def notification(cls, method: str, params: Any) -> Message:
        return cls(method=method, params=params)

    @classmethod
    def success(cls, id: Any, result: Any) -> Message:
        return cls(id=id, result=result)

    @classmethod
    def error_response(cls, id: Any | None, error: RpcError) -> Message:
        return cls(id=id, error=error)

    def is_request(self) -> bool:
        return self.id is not None and self.method is not None

    def is_notification(self) -> bool:
        return self.id is None and self.method is not None

    def is_response(self) -> bool:
        return self.method is None and self.id is not None

    def to_dict(self) -> dict[str, Any]:
        # `skip_serializing_if = Option::is_none`: absent fields stay off the wire so
        # a request never carries a `result`/`error` key.
        out: dict[str, Any] = {"jsonrpc": self.jsonrpc}
        if self.id is not None:
            out["id"] = self.id
        if self.method is not None:
            out["method"] = self.method
        if self.params is not None:
            out["params"] = self.params
        if self.result is not None:
            out["result"] = self.result
        if self.error is not None:
            out["error"] = self.error.to_dict()
        return out

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> Message:
        err = d.get("error")
        return cls(
            jsonrpc=str(d.get("jsonrpc", "")),
            id=d.get("id"),
            method=d.get("method"),
            params=d.get("params"),
            result=d.get("result"),
            error=RpcError.from_dict(err) if isinstance(err, dict) else None,
        )


# ---------------------------------------------------------------------------
# initialize
# ---------------------------------------------------------------------------


@dataclass
class HostInfo:
    name: str
    version: str

    def to_dict(self) -> dict[str, Any]:
        return {"name": self.name, "version": self.version}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> HostInfo:
        return cls(name=str(d["name"]), version=str(d["version"]))


@dataclass
class WorkspaceInfo:
    root: str
    trusted: bool

    def to_dict(self) -> dict[str, Any]:
        return {"root": self.root, "trusted": self.trusted}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> WorkspaceInfo:
        return cls(root=str(d["root"]), trusted=bool(d["trusted"]))


@dataclass
class InitializeParams:
    protocol_version: int
    host: HostInfo
    workspace: WorkspaceInfo
    mode: str
    session: Optional[dict[str, Any]] = None
    ui_capabilities: list[str] = field(default_factory=list)
    flags: dict[str, Any] = field(default_factory=dict)
    capabilities_enabled: Any | None = None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {
            "protocol_version": self.protocol_version,
            "host": self.host.to_dict(),
            "workspace": self.workspace.to_dict(),
            "mode": self.mode,
        }
        if self.session is not None:
            out["session"] = self.session
        if self.ui_capabilities:
            out["ui_capabilities"] = list(self.ui_capabilities)
        if self.flags:
            out["flags"] = dict(self.flags)
        if self.capabilities_enabled is not None:
            out["capabilities_enabled"] = self.capabilities_enabled
        return out

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> InitializeParams:
        # `protocol_version` is required — a handshake without it is malformed.
        return cls(
            protocol_version=int(d["protocol_version"]),
            host=HostInfo.from_dict(d["host"]),
            workspace=WorkspaceInfo.from_dict(d["workspace"]),
            mode=str(d["mode"]),
            session=d.get("session"),
            ui_capabilities=list(d.get("ui_capabilities", [])),
            flags=dict(d.get("flags", {})),
            capabilities_enabled=d.get("capabilities_enabled"),
        )


@dataclass
class ExtensionInfo:
    name: str
    version: str

    def to_dict(self) -> dict[str, Any]:
        return {"name": self.name, "version": self.version}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ExtensionInfo:
        return cls(name=str(d["name"]), version=str(d["version"]))


@dataclass
class ToolRegistration:
    name: str
    description: str
    parameters: dict[str, Any]
    deferred: bool = False

    def to_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "description": self.description,
            "parameters": self.parameters,
            "deferred": self.deferred,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ToolRegistration:
        return cls(
            name=str(d["name"]),
            description=str(d["description"]),
            parameters=dict(d["parameters"]),
            deferred=bool(d.get("deferred", False)),
        )


@dataclass
class CommandRegistration:
    name: str
    description: str

    def to_dict(self) -> dict[str, Any]:
        return {"name": self.name, "description": self.description}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> CommandRegistration:
        return cls(name=str(d["name"]), description=str(d["description"]))


@dataclass
class ShortcutRegistration:
    key: str
    command: str
    description: Optional[str] = None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"key": self.key, "command": self.command}
        if self.description is not None:
            out["description"] = self.description
        return out

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ShortcutRegistration:
        return cls(key=str(d["key"]), command=str(d["command"]), description=d.get("description"))


@dataclass
class Registrations:
    tools: list[ToolRegistration] = field(default_factory=list)
    commands: list[CommandRegistration] = field(default_factory=list)
    flags: list[str] = field(default_factory=list)
    shortcuts: list[ShortcutRegistration] = field(default_factory=list)
    subscriptions: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {}
        if self.tools:
            out["tools"] = [t.to_dict() for t in self.tools]
        if self.commands:
            out["commands"] = [c.to_dict() for c in self.commands]
        if self.flags:
            out["flags"] = list(self.flags)
        if self.shortcuts:
            out["shortcuts"] = [s.to_dict() for s in self.shortcuts]
        if self.subscriptions:
            out["subscriptions"] = list(self.subscriptions)
        return out

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> Registrations:
        return cls(
            tools=[ToolRegistration.from_dict(t) for t in d.get("tools", [])],
            commands=[CommandRegistration.from_dict(c) for c in d.get("commands", [])],
            flags=list(d.get("flags", [])),
            shortcuts=[ShortcutRegistration.from_dict(s) for s in d.get("shortcuts", [])],
            subscriptions=list(d.get("subscriptions", [])),
        )


@dataclass
class InitializeResult:
    protocol_version: int
    extension: ExtensionInfo
    registrations: Registrations = field(default_factory=Registrations)

    def to_dict(self) -> dict[str, Any]:
        return {
            "protocol_version": self.protocol_version,
            "extension": self.extension.to_dict(),
            "registrations": self.registrations.to_dict(),
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> InitializeResult:
        return cls(
            protocol_version=int(d["protocol_version"]),
            extension=ExtensionInfo.from_dict(d["extension"]),
            registrations=Registrations.from_dict(d.get("registrations", {})),
        )


# ---------------------------------------------------------------------------
# hook
# ---------------------------------------------------------------------------


@dataclass
class HookOutcome:
    """An extension's reply to a ``hook``. ``action`` is one of ``continue`` /
    ``block`` / ``modify``; a ``modify`` MUST carry a ``patch``. Any other action, or
    a ``modify`` without a patch, is malformed and rejected on parse."""

    action: str
    reason: Optional[str] = None
    patch: Any | None = None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"action": self.action}
        if self.action == "block" and self.reason is not None:
            out["reason"] = self.reason
        if self.action == "modify":
            out["patch"] = self.patch
        return out

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> HookOutcome:
        action = d.get("action")
        if action == "continue":
            return cls(action="continue")
        if action == "block":
            return cls(action="block", reason=d.get("reason"))
        if action == "modify":
            if "patch" not in d:
                raise ValueError("hook outcome `modify` requires a `patch`")
            return cls(action="modify", patch=d["patch"])
        raise ValueError(f"unknown hook outcome action: {action!r}")


# ---------------------------------------------------------------------------
# tool/execute + tool/update
# ---------------------------------------------------------------------------


@dataclass
class ToolExecuteParams:
    call_id: str
    tool: str
    arguments: Any
    context: Context

    def to_dict(self) -> dict[str, Any]:
        return {
            "call_id": self.call_id,
            "tool": self.tool,
            "arguments": self.arguments,
            "context": self.context.to_dict(),
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ToolExecuteParams:
        # `call_id` is required — it correlates tool/update progress + $/cancel.
        return cls(
            call_id=str(d["call_id"]),
            tool=str(d["tool"]),
            arguments=d.get("arguments"),
            context=Context.from_dict(d["context"]),
        )


@dataclass
class ToolExecuteResult:
    content: str
    is_error: bool = False
    details: Any | None = None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"content": self.content, "is_error": self.is_error}
        if self.details is not None:
            out["details"] = self.details
        return out

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ToolExecuteResult:
        # `content` is required — the LLM-facing text of the tool result.
        return cls(
            content=str(d["content"]),
            is_error=bool(d.get("is_error", False)),
            details=d.get("details"),
        )


@dataclass
class ToolUpdateParams:
    call_id: str
    message: Optional[str] = None
    progress: Optional[float] = None
    details: Any | None = None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"call_id": self.call_id}
        if self.message is not None:
            out["message"] = self.message
        if self.progress is not None:
            out["progress"] = self.progress
        if self.details is not None:
            out["details"] = self.details
        return out

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> ToolUpdateParams:
        return cls(
            call_id=str(d["call_id"]),
            message=d.get("message"),
            progress=d.get("progress"),
            details=d.get("details"),
        )


# ---------------------------------------------------------------------------
# event
# ---------------------------------------------------------------------------


@dataclass
class EventParams:
    event: str
    context: Context
    seq: Optional[int] = None
    payload: Any | None = None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"event": self.event, "context": self.context.to_dict()}
        if self.seq is not None:
            out["seq"] = self.seq
        if self.payload is not None:
            out["payload"] = self.payload
        return out

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> EventParams:
        return cls(
            event=str(d["event"]),
            context=Context.from_dict(d["context"]),
            seq=d.get("seq"),
            payload=d.get("payload"),
        )


@dataclass
class CommandExecuteResult:
    content: Optional[str] = None

    def to_dict(self) -> dict[str, Any]:
        return {"content": self.content} if self.content is not None else {}

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> CommandExecuteResult:
        return cls(content=d.get("content"))


@dataclass
class Completion:
    value: str
    description: Optional[str] = None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"value": self.value}
        if self.description is not None:
            out["description"] = self.description
        return out

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> Completion:
        return cls(value=str(d["value"]), description=d.get("description"))


@dataclass
class CommandCompleteResult:
    completions: list[Completion] = field(default_factory=list)

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> CommandCompleteResult:
        return cls(completions=[Completion.from_dict(c) for c in d.get("completions", [])])

File: rust/smooth-operator-core/src/extension/protocol.rs
//! SEP wire protocol — JSON-RPC 2.0 frames and typed method params/results.
//!
//! SEP (the Smooth Extension Protocol) is JSON-RPC 2.0 over ndjson on an
//! extension subprocess's stdio. The canonical schemas live in the
//! `smooth-operator` repo at `spec/extension/`; the types here are the Rust
//! host's view of that wire. Field names are `snake_case` to match the spec
//! exactly (Rust field names already are, so no `rename` is needed).
//!
//! A single [`Message`] type models all four JSON-RPC frame shapes (request,
//! notification, success response, error response) with the discriminating
//! fields optional — this is what lets the host parse an arbitrary inbound line
//! and classify it, and it round-trips every conformance fixture cleanly.

use serde::{Deserialize, Serialize};

/// JSON-RPC error codes. Standard range plus the SEP extensions documented in
/// `spec/extension/envelope.md`.
pub mod codes {
    pub const PARSE_ERROR: i64 = -32700;
    pub const INVALID_REQUEST: i64 = -32600;
    pub const METHOD_NOT_FOUND: i64 = -32601;
    pub const INVALID_PARAMS: i64 = -32602;
    pub const INTERNAL_ERROR: i64 = -32603;

    /// A hook or policy vetoed the operation.
    pub const BLOCKED: i64 = -32000;
    /// `ui/request` in a headless/uncapable frontend.
    pub const NO_UI: i64 = -32001;
    /// Extension acted beyond its granted trust.
    pub const NOT_TRUSTED: i64 = -32002;
    /// Command-tier action attempted from an event-tier context.
    pub const CONTEXT_VIOLATION: i64 = -32003;
    /// Method requires a capability the handshake did not enable.
    pub const CAPABILITY_DISABLED: i64 = -32004;
    /// Request cancelled via `$/cancel`.
    pub const CANCELLED: i64 = -32800;
}

/// A JSON-RPC id: an integer or a string. Modeled as a `serde_json::Value` so
/// both forms (and `null`, valid only on a parse-error response) round-trip
/// without a bespoke enum.
pub type Id = serde_json::Value;

/// A JSON-RPC error object.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RpcError {
    pub code: i64,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub data: Option<serde_json::Value>,
}

impl RpcError {
    #[must_use]
    pub fn new(code: i64, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
            data: None,
        }
    }
}

impl std::fmt::Display for RpcError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "JSON-RPC error {}: {}", self.code, self.message)
    }
}

impl std::error::Error for RpcError {}

/// The JSON-RPC 2.0 envelope. All four frame shapes share this struct; which
/// fields are present determines the shape:
///
/// - request: `id` + `method` (+ optional `params`)
/// - notification: `method`, no `id`
/// - success response: `id` + `result`
/// - error response: `id` + `error`
///
/// `skip_serializing_if` keeps absent fields off the wire so a request never
/// carries a `result`/`error` key (the spec's `additionalProperties: false`).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Message {
    pub jsonrpc: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<Id>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub method: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub params: Option<serde_json::Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub result: Option<serde_json::Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<RpcError>,
}

impl Message {
    /// Build a request frame.
    #[must_use]
    pub fn request(id: Id, method: impl Into<String>, params: serde_json::Value) -> Self {
        Self {
            jsonrpc: "2.0".into(),
            id: Some(id),
            method: Some(method.into()),
            params: Some(params),
            result: None,
            error: None,
        }
    }

    /// Build a notification frame (no id, no reply expected).
    #[must_use]
    pub fn notification(method: impl Into<String>, params: serde_json::Value) -> Self {
        Self {
            jsonrpc: "2.0".into(),
            id: None,
            method: Some(method.into()),
            params: Some(params),
            result: None,
            error: None,
        }
    }

    /// Build a success response frame echoing `id`.
    #[must_use]
    pub fn success(id: Id, result: serde_json::Value) -> Self {
        Self {
            jsonrpc: "2.0".into(),
            id: Some(id),
            method: None,
            params: None,
            result: Some(result),
            error: None,
        }
    }

    /// Build an error response frame echoing `id`.
    #[must_use]
    pub fn error_response(id: Option<Id>, error: RpcError) -> Self {
        Self {
            jsonrpc: "2.0".into(),
            id,
            method: None,
            params: None,
            result: None,
            error: Some(error),
        }
    }

    /// True when this frame is a request (has both `id` and `method`).
    #[must_use]
    pub fn is_request(&self) -> bool {
        self.id.is_some() && self.method.is_some()
    }

    /// True when this frame is a notification (has `method`, no `id`).
    #[must_use]
    pub fn is_notification(&self) -> bool {
        self.id.is_none() && self.method.is_some()
    }

    /// True when this frame is a response (has `id`, no `method`).
    #[must_use]
    pub fn is_response(&self) -> bool {
        self.method.is_none() && self.id.is_some()
    }
}

// ---------------------------------------------------------------------------
// The two-tier dispatch context.
// ---------------------------------------------------------------------------

/// Whether a dispatch may only observe (`Event`) or may mutate the session
/// (`Command`). Session-mutating ext→host actions require `Command`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Tier {
    Event,
    Command,
}

/// The dispatch context carried by every host→ext event/hook/tool/command.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Context {
    pub token: String,
    pub tier: Tier,
}

// ---------------------------------------------------------------------------
// initialize
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct HostInfo {
    pub name: String,
    pub version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct WorkspaceInfo {
    pub root: String,
    pub trusted: bool,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct SessionInfo {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InitializeParams {
    pub protocol_version: u32,
    pub host: HostInfo,
    pub workspace: WorkspaceInfo,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub session: Option<SessionInfo>,
    pub mode: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub ui_capabilities: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub capabilities_enabled: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ExtensionInfo {
    pub name: String,
    pub version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolRegistration {
    pub name: String,
    pub description: String,
    pub parameters: serde_json::Value,
    #[serde(default)]
    pub deferred: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CommandRegistration {
    pub name: String,
    pub description: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct Registrations {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub tools: Vec<ToolRegistration>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub commands: Vec<CommandRegistration>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub flags: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub subscriptions: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct InitializeResult {
    pub protocol_version: u32,
    pub extension: ExtensionInfo,
    #[serde(default)]
    pub registrations: Registrations,
}

// ---------------------------------------------------------------------------
// hook
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct HookParams {
    pub hook: String,
    pub context: Context,
    pub input: serde_json::Value,
}

/// An extension's reply to a `hook`. Serializes tagged by `action`.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "action", rename_all = "snake_case")]
pub enum HookOutcome {
    /// Proceed unchanged.
    Continue,
    /// Veto the intercepted operation.
    Block {
        #[serde(default, skip_serializing_if = "Option::is_none")]
        reason: Option<String>,
    },
    /// Replace the intercepted value with `patch`.
    Modify { patch: serde_json::Value },
}

// ---------------------------------------------------------------------------
// tool/execute + tool/update
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolExecuteParams {
    pub call_id: String,
    pub tool: String,
    pub arguments: serde_json::Value,
    pub context: Context,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolExecuteResult {
    pub content: String,
    #[serde(default)]
    pub is_error: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub details: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ToolUpdateParams {
    pub call_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub progress: Option<f64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub details: Option<serde_json::Value>,
}

// ---------------------------------------------------------------------------
// $/cancel, event, log
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CancelParams {
    pub id: Id,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct EventParams {
    pub event: String,
    /// Per-connection monotonic sequence. Absent on the out-of-band
    /// `events_lost` marker (a gap in the run is itself the loss signal).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub seq: Option<u64>,
    pub context: Context,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub payload: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct LogParams {
    pub level: String,
    pub message: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fields: Option<serde_json::Value>,
}

/// SEP method names, centralized so the host and tests never spell one wrong.
pub mod method {
    pub const INITIALIZE: &str = "initialize";
    pub const SHUTDOWN: &str = "shutdown";
    pub const PING: &str = "ping";
    pub const EVENT: &str = "event";
    pub const HOOK: &str = "hook";
    pub const TOOL_EXECUTE: &str = "tool/execute";
    pub const TOOL_UPDATE: &str = "tool/update";
    pub const COMMAND_EXECUTE: &str = "command/execute";
    pub const CANCEL: &str = "$/cancel";
    pub const REGISTRY_UPDATE: &str = "registry/update";
    pub const TOOLS_SET_ACTIVE: &str = "tools/set_active";
    pub const EXEC_RUN: &str = "exec/run";
    pub const UI_REQUEST: &str = "ui/request";
    pub const LOG: &str = "log";
    pub const BUS_PUBLISH: &str = "bus/publish";
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn roundtrip<T>(value: &T) -> T
    where
        T: Serialize + for<'de> Deserialize<'de>,
    {
        let s = serde_json::to_string(value).expect("serialize");
        serde_json::from_str(&s).expect("deserialize")
    }

    #[test]
    fn message_classification() {
        let req = Message::request(json!(1), "ping", json!({}));
        assert!(req.is_request() && !req.is_notification() && !req.is_response());

        let note = Message::notification("event", json!({}));
        assert!(note.is_notification() && !note.is_request());

        let ok = Message::success(json!(1), json!({}));
        assert!(ok.is_response() && !ok.is_request());

        let err = Message::error_response(Some(json!(1)), RpcError::new(codes::BLOCKED, "no"));
        assert!(err.is_response());
    }

    #[test]
    fn request_frame_omits_result_and_error() {
        let req = Message::request(json!(7), "tool/execute", json!({"x": 1}));
        let s = serde_json::to_string(&req).expect("serialize");
        assert!(!s.contains("result"), "{s}");
        assert!(!s.contains("error"), "{s}");
        assert!(s.contains("\"jsonrpc\":\"2.0\""));
        assert!(s.contains("\"method\":\"tool/execute\""));
    }

    #[test]
    fn notification_has_no_id() {
        // Use params without an `id` key so the top-level-id check is meaningful.
        let note = Message::notification("event", json!({"event": "turn_start"}));
        assert!(note.id.is_none());
        let s = serde_json::to_string(&note).expect("serialize");
        assert!(!s.contains("\"id\""), "{s}");
    }

    #[test]
    fn message_roundtrips_all_shapes() {
        for m in [
            Message::request(json!("abc"), "initialize", json!({})),
            Message::notification("log", json!({"level": "info", "message": "hi"})),
            Message::success(json!(2), json!({"ok": true})),
            Message::error_response(Some(json!(2)), RpcError::new(codes::CANCELLED, "cancelled")),
            Message::error_response(None, RpcError::new(codes::PARSE_ERROR, "bad json")),
        ] {
            assert_eq!(roundtrip(&m), m);
        }
    }

    #[test]
    fn initialize_params_roundtrip() {
        let p = InitializeParams {
            protocol_version: 1,
            host: HostInfo {
                name: "smooth-operator-core".into(),
                version: "0.15.0".into(),
            },
            workspace: WorkspaceInfo {
                root: "/ws".into(),
                trusted: true,
            },
            session: Some(SessionInfo { id: Some("s1".into()) }),
            mode: "headless".into(),
            ui_capabilities: vec!["confirm".into()],
            capabilities_enabled: Some(json!({"tools": true})),
        };
        assert_eq!(roundtrip(&p), p);
    }

    #[test]
    fn initialize_result_roundtrip() {
        let r = InitializeResult {
            protocol_version: 1,
            extension: ExtensionInfo {
                name: "echo".into(),
                version: "0.1.0".into(),
            },
            registrations: Registrations {
                tools: vec![ToolRegistration {
                    name: "say".into(),
                    description: "Echo a phrase back.".into(),
                    parameters: json!({"type": "object"}),
                    deferred: false,
                }],
                subscriptions: vec!["turn_start".into()],
                ..Default::default()
            },
        };
        assert_eq!(roundtrip(&r), r);
    }

    #[test]
    fn hook_outcome_variants_serialize_by_action() {
        assert_eq!(serde_json::to_value(HookOutcome::Continue).unwrap(), json!({"action": "continue"}));
        assert_eq!(
            serde_json::to_value(HookOutcome::Block { reason: Some("nope".into()) }).unwrap(),
            json!({"action": "block", "reason": "nope"})
        );
        assert_eq!(serde_json::to_value(HookOutcome::Block { reason: None }).unwrap(), json!({"action": "block"}));
        assert_eq!(
            serde_json::to_value(HookOutcome::Modify { patch: json!({"a": 1}) }).unwrap(),
            json!({"action": "modify", "patch": {"a": 1}})
        );
    }

    #[test]
    fn hook_outcome_parses_from_wire() {
        let c: HookOutcome = serde_json::from_value(json!({"action": "continue"})).unwrap();
        assert_eq!(c, HookOutcome::Continue);
        let m: HookOutcome = serde_json::from_value(json!({"action": "modify", "patch": {}})).unwrap();
        assert!(matches!(m, HookOutcome::Modify { .. }));
    }

    #[test]
    fn hook_outcome_rejects_unknown_action() {
        let r: Result<HookOutcome, _> = serde_json::from_value(json!({"action": "bogus"}));
        assert!(r.is_err());
    }

    #[test]
    fn tier_serializes_snake_case() {
        assert_eq!(serde_json::to_value(Tier::Command).unwrap(), json!("command"));
        assert_eq!(serde_json::to_value(Tier::Event).unwrap(), json!("event"));
    }

    #[test]
    fn tool_execute_roundtrip() {
        let p = ToolExecuteParams {
            call_id: "c1".into(),
            tool: "say".into(),
            arguments: json!({"phrase": "hi"}),
            context: Context {
                token: "t".into(),
                tier: Tier::Command,
            },
        };
        assert_eq!(roundtrip(&p), p);
        let r = ToolExecuteResult {
            content: "hi".into(),
            is_error: false,
            details: None,
        };
        assert_eq!(roundtrip(&r), r);
    }

    #[test]
    fn rpc_error_is_std_error() {
        let e = RpcError::new(codes::NO_UI, "headless");
        assert_eq!(e.to_string(), "JSON-RPC error -32001: headless");
    }
}
